File: scripts/algorithms/M/Median of Two Sorted Arrays/Median of Two Sorted Arrays.js
// Runtime: 79 ms (Top 92.6%) | Memory: 46.70 MB (Top 75.57%)

var findMedianSortedArrays = function(nums1, nums2) {
    if (nums1.length > nums2.length) {
        [nums1, nums2] = [nums2, nums1];
    }
    
    const m = nums1.length;
    const n = nums2.length;
    let low = 0, high = m;
    
    while (low <= high) {
        const partitionX = Math.floor((low + high) / 2);
        const partitionY = Math.floor((m + n + 1) / 2) - partitionX;
        
        const maxX = (partitionX === 0) ? Number.MIN_SAFE_INTEGER : nums1[partitionX - 1];
        const maxY = (partitionY === 0) ? Number.MIN_SAFE_INTEGER : nums2[partitionY - 1];
        
        const minX = (partitionX === m) ? Number.MAX_SAFE_INTEGER : nums1[partitionX];
        const minY = (partitionY === n) ? Number.MAX_SAFE_INTEGER : nums2[partitionY];
        
        if (maxX <= minY && maxY <= minX) {
            if ((m + n) % 2 === 0) {
                return (Math.max(maxX, maxY) + Math.min(minX, minY)) / 2;
            } else {
                return Math.max(maxX, maxY);
            }
        } else if (maxX > minY) {
            high = partitionX - 1;
        } else {
            low = partitionX + 1;
        }
    }
    
    throw new Error("Input arrays are not sorted.");
};

File: scripts/algorithms/M/Median of Two Sorted Arrays/Median of Two Sorted Arrays.java
// Runtime: 5 ms (Top 46.02%) | Memory: 50.6 MB (Top 18.35%)

class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if(nums1.length==1 && nums2.length==1) return (double)(nums1[0]+nums2[0])/2.0;
        int i = 0;
        int j = 0;
        int k = 0;
        int nums[] = new int[nums1.length + nums2.length];
        if(nums1.length !=0 && nums2.length != 0){
          while(i < nums1.length && j < nums2.length){
            if(nums1[i] < nums2[j]){
                nums[k] = nums1[i];
                i++;
                k++;
            }else{
                nums[k] = nums2[j];
                j++;
                k++;
            }
          }
            while(i < nums1.length){
                nums[k] = nums1[i];
                i++;
                k++;
            }
             while(j < nums2.length){
                nums[k] = nums2[j];
                j++;
                k++;
            }
        }
        if(nums1.length==0){
            for(int h: nums2){
                nums[k] = h;
                k++;
            }
        }

        if(nums2.length==0){
            for(int d : nums1){
                nums[k] = d;
                k++;
            }
        }

        int mid = (nums.length / 2);

        if (nums.length % 2 == 0) {
            return ((double) nums[mid] + (double) nums[mid - 1]) / 2.0;
        } else {
            return nums[mid];
        }
    }
}